src/core/Song.cpp

```diff
@@ -191,226 +193,159 @@ void Song::savePos()
-		m_vstSyncController.setPlaybackJumped( true );
-		m_playPos[m_playMode].setJumped( false );
+		m_vstSyncController.setPlaybackJumped(true);
+		getPlayPos().setJumped(false);
 	}
 
-	f_cnt_t framesPlayed = 0;
-	const float framesPerTick = Engine::framesPerTick();
+	const auto framesPerTick = Engine::framesPerTick();
+	const auto framesPerPeriod = Engine::mixer()->framesPerPeriod();
+
+	f_cnt_t frameOffsetInPeriod = 0;
 
-	while( framesPlayed < Engine::mixer()->framesPerPeriod() )
+	while (frameOffsetInPeriod < framesPerPeriod)
 	{
-		m_vstSyncController.update();
+		auto frameOffsetInTick = getPlayPos().currentFrame();
 
-		float currentFrame = m_playPos[m_playMode].currentFrame();
-		// did we play a tick?
-		if( currentFrame >= framesPerTick )
+		// If a whole tick has elapsed, update the frame and tick count, and check any loops
+		if (frameOffsetInTick >= framesPerTick)
 		{
-			int ticks = m_playPos[m_playMode].getTicks() +
-				( int )( currentFrame / framesPerTick );
-
-			// did we play a whole bar?
-			if( ticks >= MidiTime::ticksPerBar() )
+			// Transfer any whole ticks from the frame count to the tick count
+			const auto elapsedTicks = static_cast<int>(frameOffsetInTick / framesPerTick);
+			getPlayPos().setTicks(getPlayPos().getTicks() + elapsedTicks);
+			frameOffsetInTick -= elapsedTicks * framesPerTick;
+			getPlayPos().setCurrentFrame(frameOffsetInTick);
+
+			// If we are playing a BB track, or a pattern with no loop enabled,
+			// loop back to the beginning when we reach the end
+			if (m_playMode == Mode_PlayBB)
 			{
-				// per default we just continue playing even if
-				// there's no more stuff to play
-				// (song-play-mode)
-				int maxBar = m_playPos[m_playMode].getBar()
-									+ 2;
-
-				// then decide whether to go over to next bar
-				// or to loop back to first bar
-				if( m_playMode == Mode_PlayBB )
-				{
-					maxBar = Engine::getBBTrackContainer()
-							->lengthOfCurrentBB();
-				}
-				else if( m_playMode == Mode_PlayPattern &&
-					m_loopPattern == true &&
-					tl != NULL &&
-					tl->loopPointsEnabled() == false )
-				{
-					maxBar = m_patternToPlay->length()
-								.getBar();
-				}
-
-				// end of played object reached?
-				if( m_playPos[m_playMode].getBar() + 1
-								>= maxBar )
-				{
-					// then start from beginning and keep
-					// offset
-					ticks %= ( maxBar * MidiTime::ticksPerBar() );
-
-					// wrap milli second counter
-					setToTimeByTicks(ticks);
-
-					m_vstSyncController.setPlaybackJumped( true );
-				}
+				enforceLoop(MidiTime{0}, MidiTime{Engine::getBBTrackContainer()->lengthOfCurrentBB(), 0});
+			}
+			else if (m_playMode == Mode_PlayPattern && m_loopPattern && !loopEnabled)
+			{
+				enforceLoop(MidiTime{0}, m_patternToPlay->length());
 			}
-			m_playPos[m_playMode].setTicks( ticks );
 
-			if (checkLoop || m_loopRenderRemaining > 1)
+			// Handle loop points, and inform VST plugins of the loop status
+			if (loopEnabled || m_loopRenderRemaining > 1)
```

Comment:
```suggestion
			if (loopEnabled || (m_loopRenderRemaining > 1 && getPlayPos() >= timeline->loopBegin()))
```
This will be explained in the comment below.